include/swift/AST/LocalizationFormat.h

```diff
@@ -181,16 +184,35 @@ class SerializedLocalizationProducer final : public LocalizationProducer {
 class LocalizationInput : public llvm::yaml::Input {
   using Input::Input;
 
+  /// A diagnostic ID might be present in YAML and not in `.def` file, if that's
+  /// the case the `id` won't have a `DiagID` value, so the current diagnostic
+  /// ID will be read as a string instead. Since the diagnostic ID isn't
+  /// available in `.def` it shouldn't be stored in the diagnostics array, but
+  /// rather stored in `unknownIDs`.
+  ///
+  /// \param io The `llvm::yaml::IO`
+  /// \param id The current diagnostic ID. It's supposed to be an `uint32_t`.
+  /// \param unknownIDs An array that keeps track of all of diagnostic IDs that
+  /// are available in YAML and not available in `.def` files. \returns true if
+  /// the current `id` is not avialiable in `.def`, false otherwise.
```

Comment:
@xedin WDYT about this documentation?

Reply:
Looks good besides `\param` section:

`io` - needs a description instead of class name `llvm::yaml::IO`.
`id` - it needs to be explain that id is what gets read from `io` and stored.

I'd prefer it if we didn't have side-effects in this method. It would be better if `readID` returned `llvm::Optional<uint32_t>` instead of taking a reference as a parameter and if that's `None` - `readYAML` would execute fallback logic and read `id` as a string and store it into `unknownIDs` directly (it would be great if we didn't have to thread that all over the place).